src/lib/tab_groups.ts

```diff
@@ -0,0 +1,330 @@
+import { browserBg, activeTabId } from "./webext"
+
+/**
+ * Return the active window's id.
+ *
+ */
+export async function activeWindowId() {
+    return (await browserBg.windows.getCurrent()).id
+}
+
+/**
+ * Return a set of the current window's tab groups (empty if there are none).
+ *
+ */
+export async function tgroups() {
+    const groups = await browserBg.sessions.getWindowValue(await activeWindowId(), "tridactyl-tgroups")
+    return new Set<string>(groups as string[])
+}
+
+/**
+ * Set the current window's tab groups.
+ *
+ * @param groups The set of groups.
+ *
+ */
+export async function setTgroups(groups: Set<string>) {
+    return browserBg.sessions.setWindowValue(await activeWindowId(), "tridactyl-tgroups", [...groups])
+}
+
+/**
+ * Clear the current window's tab groups.
+ *
+ */
+export async function clearTgroups() {
+    return browserBg.sessions.removeWindowValue(await activeWindowId(), "tridactyl-tgroups")
+}
+
+/**
+ * Return the active tab group for the a window or undefined.
+ *
+ * @param id The id of the window. Use the current window if not specified.
+ *
+ */
+export async function windowTgroup(id?: number) {
+    return browserBg.sessions.getWindowValue(id || await activeWindowId(), "tridactyl-active-tgroup")
+}
+
+/**
+ * Set the active tab group for the current window.
+ *
+ * @param name The name of the new active tab group.
+ *
+ */
+export async function setWindowTgroup(name: string) {
+    return browserBg.sessions.setWindowValue(await activeWindowId(), "tridactyl-active-tgroup", name)
+}
+
+/**
+ * Clear the active tab group for the current window.
+ *
+ */
+export async function clearWindowTgroup() {
+    return browserBg.sessions.removeWindowValue(await activeWindowId(), "tridactyl-active-tgroup")
+}
+
+/**
+ * Return a tab's tab group.
+ *
+ * @param id The id of the tab. Use the current tab if not specified.
+ *
+ */
+export async function tabTgroup(id?: number) {
+    return browserBg.sessions.getTabValue(id || await activeTabId(), "tridactyl-tgroup")
```

Comment:
I'm not seeing any restrictions on tab IDs on MDN so I think `id` could be `0` here, which would result in unexpected behavior.

Reply:
I always forget 0 is falsey in some languages. Even if it couldn't be 0, I think an explicit check is better; updated.